account_invoice_start_end_dates/models/account_move.py

```diff
@@ -0,0 +1,28 @@
+# Copyright 2019-2021 Akretion France (https://akretion.com/)
+# @author: Alexis de Lattre <alexis.delattre@akretion.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import models
+from odoo.exceptions import UserError
+
+
+class AccountMove(models.Model):
+    _inherit = "account.move"
+
+    def _post(self, soft=True):
+        for move in self:
+            missing_date_lines = move.line_ids.filtered(
+                lambda line: line.product_id
+                and line.must_have_dates
+                and (not line.start_date or not line.end_date)
+            )
+            if missing_date_lines:
+                product_name = missing_date_lines[0].product_id.display_name
+                raise UserError(
+                    self.env._(
+                        "Missing Start Date and End Date for invoice "
+                        "line with Product '{product_name}' which has the "
+                        "property 'Must Have Start/End Dates'.",
+                    ).format(product_name=product_name)
+                )
```

Comment:
late remark about this change: what pushed you to change this loop structure?
because your new code will be slower, as it will go through the exhaustive list of `move.line_ids` while the old code was exiting at first missing date case

![image](https://github.com/user-attachments/assets/7aed96b3-0774-43d6-a41d-c540656fdbb6)

Reply:
i'm not sure, the old code that loops and checks the condition might be better if the condition is true on the first items and worse if the condition is true on the last item, in fact using `for move.line_ids` requires calling all the `account.move.line` records that belong to the `move` which can be expensive if the record set is large. maybe the quote Use `filtered() helps to optimize and simplify the code by avoiding unwanted loop conditions to reach the condition` in "[here](https://www.cybrosys.com/blog/recordsets-and-operations-odoo)" made me misunderstand that using `filtered` has more advantages than traditional loop and condition statements. so i'll restore the old code, thanks for your question.